### PR TITLE
test(scanner): raise git_secrets.py coverage from 49% to 99%

### LINKED ISSUE #269: Increase test coverage for scanner/git_secrets.py (currently 49%)
## Summary

`src/envdrift/scanner/git_secrets.py` is the worst-covered module in the codebase at **49%** (146 of 313 statements uncovered), well below the project's ≥80% line-coverage bar and the 85% repo-wide average. This issue tracks raising it to ≥80%.

## Current state

Measured via `pytest --cov=envdrift`:

| Metric | Value |
|--------|-------|
| Coverage | 49% |
| Statements missed | 146 / 313 |
| Branches | 120 (9 partial) |

## Uncovered areas

Mapping the missing line ranges to functions:

- **`GitSecretsInstaller` installation paths** (`134-238`) — `install()`, `_install_homebrew()`, and `_install_from_source()` are almost entirely untested (Homebrew on macOS, `make install` from source on Linux).
- **`GitSecretsScanner.scan()`** (`422-534`) — the core scan method, the largest single gap.
- **`_find_binary()` / `install()`** (`328-329`, `338-343`).
- **Hook & pattern helpers** — `install_hooks()` (`747-755`), `add_pattern()` (`768-778`), `list_patterns()` (`789-819`).
- Minor branch gaps in `_parse_output` (`570`, `631->635`, `702`).

## Suggested approach

- Mock `subprocess.run` / `shutil.which` to exercise `scan()`, `_run_git_secrets()`, and the installer branches without invoking real `git-secrets` or Homebrew.
- Cover both success and failure/exception paths (`GitSecretsNotFoundError`, `GitSecretsInstallError`, `GitSecretsError`).
- Add `_parse_output` fixtures from representative `git-secrets` output to close the parsing branch gaps.
- Test `install_hooks`, `add_pattern` (allowed vs prohibited), and `list_patterns` against a temp repo with mocked git calls.

## Acceptance criteria

- [ ] `scanner/git_secrets.py` line coverage ≥ 80%
- [ ] New tests cover `scan()`, both installer paths, and the hook/pattern helpers
- [ ] Both success and error/exception branches exercised

### PR DESCRIPTION
**Pull request by jainal09**
## Summary

- Raises `scanner/git_secrets.py` line coverage from **49% → 99%** (was the lowest-covered module at 313 statements, 146 missed)
- Adds **52 new unit tests** across 9 new test classes in `tests/unit/test_scanner_git_tools.py`
- All 86 tests pass; only 3 lines remain uncovered (the `except ValueError: continue` defensive branch inside the git-root walk, which is practically unreachable)
- Closes #269

## Covered areas

| New test class | Lines/branches covered |
|---|---|
| `TestGitSecretsInstallerPaths` | `_install_homebrew` (already-installed, failure modes, timeout, SubprocessError); `_install_from_source` (no git/make, success, system-path fallback, CalledProcessError, timeout); unsupported platform |
| `TestGitSecretsScannerFindBinaryExtra` | git subcommand detection (lines 328-329), auto-install success/failure (lines 338-343) |
| `TestGitSecretsScan` | Full `scan()` body (lines 422-534): nonexistent paths, file/dir scan targets, `--scan-history`, timeout, non-fatal exception continuation, parent git-root walk, no-git-repo scan, AWS pattern dedup/skip/disable, unique stderr findings |
| `TestGitSecretsParseOutputBranches` | `[ERROR]`/`error:` skip (line 570), absolute-path passthrough (branch 631→635), unmatched lines |
| `TestDetectRuleTypeAWSSecretKey` | `aws_secret`/`secret_access_key` → `aws-secret-key` (line 702) |
| `TestInstallHooks` | No `.git` early-return, success, non-zero, exception (lines 747-755) |
| `TestAddPattern` | Prohibited and allowed patterns, non-zero, exception (lines 768-778) |
| `TestListPatterns` | All three categories, empty output, non-zero, exception, value-less lines (lines 789-819) |
| `TestGitSecretsScanExtra` | Directory with parent git root (branch 453→464), unique stderr append (line 520), `--register-aws` exception silently ignored (lines 492-493) |

## Test plan

- [x] `pytest tests/unit/test_scanner_git_tools.py --cov=envdrift.scanner.git_secrets` → 99%, 86 passed
- [x] All pre-commit hooks pass (ruff, pyrefly, bandit)

🤖 Generated with [Claude Code](https://claude.ai/claude-code)

<h3>Greptile Summary</h3>

Raises line coverage of `scanner/git_secrets.py` from 49 % to 99 % by adding 52 new unit tests across 9 test classes, all of which pass. No production code is modified.

- **Installation paths** (`TestGitSecretsInstallerPaths`): covers Homebrew already-installed detection, timeout/SubprocessError/CalledProcessError handling, source-build fallbacks, and unsupported-platform rejection.
- **`scan()` body** (`TestGitSecretsScan`, `TestGitSecretsScanExtra`): covers nonexistent path skipping, file vs. directory targets, `--scan-history` flag, AWS pattern deduplication, unique-stderr appending, `--register-aws` exception swallowing, non-fatal exception loop continuation, and parent git-root walking.
- **`_parse_output` / `_detect_rule_type` / `install_hooks` / `add_pattern` / `list_patterns`**: each branch and edge case is verified with focused tests and precise assertions.

<h3>Confidence Score: 5/5</h3>

Test-only PR with no production code changes; all 52 new tests correctly model the production scan() logic and are safe to merge.

Every mock sequence was cross-checked against the actual scan() implementation: call ordering for --list/--register-aws/--scan, the total_files counter incremented before the try block, parent git-root walking, TimeoutExpired early return, and stderr deduplication via the frozen-dataclass __eq__. No logic errors or incorrect assertions were found.

No files require special attention.

<h3>Important Files Changed</h3>

| Filename | Overview |
|----------|----------|
| tests/unit/test_scanner_git_tools.py | Adds 52 new unit tests in 9 classes; mock sequences, path assertions, and exception-continuation checks are all consistent with the production scan() logic. |

</details>

<h3>Flowchart</h3>

```mermaid
%%{init: {'theme': 'neutral'}}%%
flowchart TD
    A[scan called] --> B{path.exists?}
    B -- No --> C[continue — skip]
    B -- Yes --> D{path.is_file?}
    D -- Yes --> E[total_files++\ncwd = parent]
    D -- No --> F[cwd = path]
    E & F --> G{.git in cwd?}
    G -- No --> H[Walk parents for .git]
    H -- Found --> I{is_file?\nrelative_to git_root}
    I -- ValueError --> J[continue — outside repo]
    I -- OK --> K[update scan_paths]
    H -- Not found --> L[cwd stays as-is]
    G -- Yes --> M[cwd = git_root]
    K & L & M --> N{.git exists at cwd?}
    N -- Yes --> O[--list]
    O --> P{register_aws AND\nno AWS patterns?}
    P -- Yes --> Q[--register-aws\nexcept log & continue]
    P -- No --> R
    Q --> R{include_git_history?}
    N -- No --> R
    R -- Yes --> S[--scan-history]
    R -- No --> T[--scan -r paths]
    S & T --> U{TimeoutExpired?}
    U -- Yes --> V[return ScanResult with error]
    U -- No --> W{returncode != 0\nAND stdout?}
    W -- Yes --> X[parse stdout findings]
    W -- No --> Y
    X --> Y{stderr non-empty?}
    Y -- Yes --> Z[parse stderr, dedup append]
    Y -- No --> AA{Exception?}
    Z --> AA
    AA -- Yes --> AB[log, continue loop]
    AA -- No --> AC[next path]
    AC --> B
    AB --> B
```

<sub>Reviews (4): Last reviewed commit: ["Merge branch &#39;main&#39; into feat/git-secret..."](https://github.com/jainal09/envdrift/commit/93d99a7fc84fb4f2fefced1929fb3dc1e671c8bf) | [Re-trigger Greptile](https://app.greptile.com/api/retrigger?id=34610927)</sub>

### CODE REVIEW

**Comment by jainal09** [comment]
Reviewed; no blocking findings from the diff. The added tests are test-only changes around `git_secrets.py`, and the focused suite passes locally.\n\nValidation run locally on the PR worktree:\n- `uv run --extra dev pytest tests/unit/test_scanner_git_tools.py -q`

**Comment by jainal09** [comment]
@coderabbitai review

File: tests/unit/test_scanner_git_tools.py [hunk 1]
```diff
@@ -300,3 +305,754 @@ def test_config_all_scanners_disabled(self) -> None:
...
+            mock_run.return_value = MagicMock(returncode=1, stdout="", stderr="Some brew error")
+            installer = GitSecretsInstaller()
+            with pytest.raises(GitSecretsInstallError, match="Homebrew installation failed"):
+                installer._install_homebrew()
+
+    def test_install_homebrew_binary_not_found_after_install_raises(self) -> None:
+        """Test _install_homebrew raises when git-secrets is absent after install."""
+        with patch("shutil.which") as mock_which, patch("subprocess.run") as mock_run:
+            mock_which.side_effect = lambda x: ("/usr/local/bin/brew" if x == "brew" else None)
+            mock_run.return_value = MagicMock(returncode=0, stdout="", stderr="")
```
[the remark this answers is not in the document]
Comment (jainal09) [reply tests/unit/test_scanner_git_tools.py:394]: Fixed in ace4156 — added `assert result.files_scanned == 2` alongside the existing `call_count == 6` check. The `files_scanned` counter is incremented before the `try` block, so it directly proves the loop reached and counted path2 after the exception on path1.

**Comment by jainal09** [comment]
@coderabbitai review
